sky/backends/cloud_vm_ray_backend.py

```diff
@@ -1656,6 +1657,28 @@ def provision_with_retries(
         launchable_retries_disabled = (self._dag is None or
                                        self._optimize_target is None)
 
+        # Re-optimize if to_provision.cloud does not support requested features
+        while not to_provision.cloud.support(self._requested_features):
+            if launchable_retries_disabled:
+                with ux_utils.print_exception_no_traceback():
+                    raise exceptions.NotSupportedError(
+                        f'{to_provision.cloud} does not support: '
+                        f'{self._requested_features}. DAG and '
+                        'optimize_target needs to be registered first '
+                        'to enable cross-cloud retry.')
+            self._blocked_resources.add(
+                resources_lib.Resources(cloud=to_provision.cloud))
+            # Set to None so that sky.optimize() will assign a new one
+            # (otherwise will skip re-optimizing this task).
+            # TODO: set all remaining tasks' best_resources to None.
+            task.best_resources = None
+            self._dag = sky.optimize(self._dag,
+                                     minimize=self._optimize_target,
+                                     blocked_resources=self._blocked_resources)
+            to_provision = task.best_resources
+            assert task in self._dag.tasks, 'Internal logic error.'
+            assert to_provision is not None, task
```

Comment:
Instead of having the loop here, how about we add the following line before L1694, so we can reuse the re-optimization logic there?
```
if not to_provision.cloud.support(self._requested_features):
    self._blokced_resources.add(resources_lib.Resources(cloud=to_provision.cloud)
    raise ResourcesUnavailableError('f'{to_provision.cloud} does not support: '
                        f'{self._requested_features}. )
```

Reply:
I like this idea; adding the loop did not feel good.

Reply:
Hmm, in this case adding to `blocked_resources` might not be needed at all

Reply:
Implemented! Looks much nicer than what I had originally :)